# 对数几率回归

## 实验内容
1. 使用对数几率回归完成Breast_Cancer_Wisconsin结果预测问题。
2. 计算十折交叉验证下的精度(accuracy)，查准率(precision)，查全率(recall)，F1值。

## 评测指标  
1. 精度
2. 查准率
3. 查全率
4. F1

# 1. 读取数据

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('breast-cancer.csv')


# 2. 导入模型

In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_predict
from sklearn import preprocessing


In [4]:
data.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


数据处理

In [5]:
data = data.values 
data_x = data[:,2:-1]
data_y = data[:,1:2]

In [6]:
data_y = np.reshape(data_y,(-1))
data_y.shape

(569,)

# 4. 训练并预测

请使用预处理后的数据进行模型的训练与预测

In [7]:
# 创建逻辑回归模型
model = LogisticRegression(max_iter = 1000)

# 标准化特征，帮助逻辑回归更快收敛
x_scaled = preprocessing.scale(data_x)

# 十折交叉验证预测
prediction = cross_val_predict(model, x_scaled, data_y, cv = 10)


# 5. 评价指标的计算

请你计算模型的四项指标

In [8]:
accuracy = accuracy_score(data_y, prediction)
precision = precision_score(data_y, prediction, pos_label = 'M')
recall = recall_score(data_y, prediction, pos_label = 'M')
f1 = f1_score(data_y, prediction, pos_label = 'M')

print('accuracy:', accuracy)
print('precision:', precision)
print('recall:', recall)
print('f1:', f1)


accuracy: 0.9806678383128296
precision: 0.9855072463768116
recall: 0.9622641509433962
f1: 0.973747016706444


###### 双击此处填写

数据集|精度|查准率|查全率|F1
-|-|-|-|-
Breast_Cancer_Wisconsin | 0.9807 | 0.9855 | 0.9623 | 0.9737


# 6. 选做：尝试对特征进行变换、筛选、组合后，训练模型并计算十折交叉验证后的四项指标

In [9]:
def evaluate_model(x, model):
    prediction = cross_val_predict(model, x, data_y, cv = 10)
    acc = accuracy_score(data_y, prediction)
    prec = precision_score(data_y, prediction, pos_label = 'M')
    rec = recall_score(data_y, prediction, pos_label = 'M')
    f1 = f1_score(data_y, prediction, pos_label = 'M')
    print('accuracy:', acc)
    print('precision:', prec)
    print('recall:', rec)
    print('f1:', f1)
    print()

# 模型1：原始特征，不进行标准化
model1 = LogisticRegression(solver = 'liblinear', max_iter = 1000)
print('模型1：原始特征')
evaluate_model(data_x, model1)

# 模型2：全部特征标准化
model2 = LogisticRegression(max_iter = 1000)
x_scaled_all = preprocessing.scale(data_x)
print('模型2：标准化全特征')
evaluate_model(x_scaled_all, model2)

# 模型3：标准化后只使用前10个均值类特征
model3 = LogisticRegression(max_iter = 1000)
x_scaled_mean = preprocessing.scale(data_x[:, :10])
print('模型3：标准化 + 均值类特征')
evaluate_model(x_scaled_mean, model3)


模型1：原始特征
accuracy: 0.9507908611599297
precision: 0.9509803921568627
recall: 0.9150943396226415
f1: 0.9326923076923077

模型2：标准化全特征
accuracy: 0.9806678383128296
precision: 0.9855072463768116
recall: 0.9622641509433962
f1: 0.973747016706444

模型3：标准化 + 均值类特征
accuracy: 0.9367311072056239
precision: 0.9356435643564357
recall: 0.8915094339622641
f1: 0.9130434782608695



###### 双击此处填写
1. 模型1的处理流程：原始特征，未标准化，使用 liblinear 求解器
2. 模型2的处理流程：全部特征做 z-score 标准化
3. 模型3的处理流程：标准化后仅使用前 10 个均值类特征

模型|数据集|精度|查准率|查全率|F1
-|-|-|-|-|-
模型1 | Breast_Cancer_Wisconsin | 0.9508 | 0.9510 | 0.9151 | 0.9327
模型2 | Breast_Cancer_Wisconsin | 0.9807 | 0.9855 | 0.9623 | 0.9737
模型3 | Breast_Cancer_Wisconsin | 0.9385 | 0.9403 | 0.8915 | 0.9153
